Clearly separate compatibility rules for input and output schemas. (#851)
zalando/restful-api-guidelines: chapters/compatibility.adoc

== {SHOULD} prefer compatible extensions

API designers should apply the following rules to evolve RESTful APIs for

> services in a backward-compatible way.

Reviewer: ```suggestion
services in a compatible way.
```

In general:

* Never change the semantic of fields (e.g. changing the semantic from

>   customer-number to customer-id, as both are different unique customer keys)

Reviewer: ```suggestion
  customer-number to customer-id, both being unique, but different customer keys)
```

* Consider <<251>> in case a URL has to change.

For schemas used in input only:

* Add only optional, never mandatory fields (and don't make optional
  fields mandatory).

> * Don't remove fields.

Author: Just as a note for the future: If you do two related changes in adjacent lines (which don't make sense individually), please just suggest them together. Makes it easier to review.

   * While removing fields by itself doesn't break compatibility (if the
     server would still accept it, possibly ignoring it, if sent by the client),
     it then would allow later to add a same-named field with different type
     or semantic (which is harder to catch). Therefore, this is also considered
     a non-compatible change.
* Input fields may have (complex) constraints being validated via server-side
  business logic. Never change the validation logic to be more restrictive and
  make sure that all constraints are clearly defined in description.

> * `enum` ranges can be reduced when used as input, only if the server

Reviewer: ```suggestion
* `enum` ranges can be reduced, only if the server
```

>   is ready to accept and handle old range values too.

Reviewer: ```suggestion
  continues to accept and handle old range values.
```

* `enum` ranges can be extended when used for input.

> 

Reviewer: ```suggestion

(*) Hint: Removing a field can be considered as a compatible change that does not break clients, in case the service still accepts and possibly ignores it if sent by the client. However, removed fields allow later adding a same-named field with different type or semantic (which is harder to catch). We therefore define field removal as non-compatible.  

```

> For schemas used in output only:

Reviewer: ```suggestion
For schemas only used for output objects:
```

> * New mandatory fields can be added, or existing optional ones be made mandatory.

Reviewer: ```suggestion
* Add (mandatory or optional) fields.
* Make optional fields mandatory, but not vice-versa.
```

* Don't remove a mandatory field, or make it optional (clients might depend
  on it being filled).
* Don't remove optional fields either.
   * While removing optional fields by itself doesn't break compatibility,
     it then would allow later to add a same-named field with different type

>      or semantic (which is harder to catch). Therefore, this is also considered a non-compatible change.

Reviewer: ```suggestion
```

* `enum` ranges can be reduced when used as output.

> * `enum` ranges **cannot** be extended when used for output — clients may

Reviewer: ```suggestion
* `enum` ranges *cannot* be extended — clients may
```

  not be prepared to handle it.
* You <<112>> that are used for output parameters and likely to
  be extended with growing functionality. The API definition should be updated
  first before returning new values.

For schemas used in both input and output (which is common and recommended in
many cases), both of these rule sets combine, i.e. you can only do changes which
are allowed in both input and output.

* Add only optional, never mandatory fields.
* Don't remove any fields.
* Don't make mandatory fields optional or make optional fields mandatory.
* Input fields may have (complex) constraints being validated via server-side
  business logic. Never change the validation logic to be more restrictive and
  make sure that all constraints are clearly defined in description.
* `enum` ranges can be reduced only if the server is ready to still accept and
   handle old values. They **cannot** be extended.
* You <<112>> that are used for output parameters and likely to
  be extended with growing functionality. The API definition should be updated

>   first before returning new values.

Reviewer: ```suggestion
```

Reviewer: This is is a consequence of the sentence before, and redundant -- should be omitted.

Reviewer: Let's keep it shorter in this situation. The info is clear in the context, and repetition is more confusing than helpful, because you expect deviations which is not intended. BTW: This was also feedback in the Guild Meeting where we discussed the PR during your absense. I took the task to update the PR accordingly.

Input/Output here is from the perspective of a service implementing and

> owning the API. For the rare case of APIs implemented by other services

Reviewer: ```suggestion
owning the API. For the rare case of call-back APIs implemented by other services
```
